proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,200 @@
+block data distribution, so that we can lower the possibility of human mistakes
+in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages](https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc)
+for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## Detailed Design
+
+The new protocol tries to make all 80% restarting validators get the same
+data blocks and the same set of last votes among them, then they can probably
+make the same decision and then proceed.
+
+The steps roughly look like this:
+
+1. Everyone freezes, no new blocks, no new votes, and no Turbine
+
+2. Make all blocks which can potentially have been optimistically confirmed
+before the freeze propagate to everyone
+
+3. Make restart participants' last votes before the freeze propagate to
+everyone
+
+4. Now see if enough people can optimistically agree on one block (same slot
+and hash) to restart from
+
+4.1 If yes, proceed and restart
+
+4.2 If no, freeze and print out what you think is wrong, wait for human
+
+A new command line arg --RepairAndRestart is added. When the cluster is in need
+of a restart, we assume at least 80% will restart with this arg. Any validators
+restarted with this arg does not participate in the normal Turbine protocol,
+update its vote, or generate new blocks until all of the following steps are
+completed.
+
+### Gossip last vote before the restart and ancestors on that fork
+
+Send Gossip message LastVotedForkSlots to everyone in restart, it contains the
+last voted slot on its tower and the ancestor slots on the last voted fork and
+is sent in a compressed bitmap like the EpicSlots data structure. The number of
+ancestor slots sent is hard coded at 81000, because that's 400ms * 81000 = 9
+hours, we assume most restart decisions to be made in 9 hours. If a validator
+restarts after 9 hours past the outage, it cannot join the restart this way. If
+enough validators failed to restart within 9 hours, then use the old restart
+method.
+
+The fields of LastVotedForkSlots are:
+
+- `last_voted_slot`: the slot last voted, this also serves as last_slot for the
+bit vector.
+- `last_voted_hash`: the bank hash of the slot last voted slot.
+- `slots`: compressed bit vector representing the slots on the last voted fork,
+last slot is always last_voted_slot, first slot is last_voted_slot-81000.
+
+When a validator enters restart, it increments its current shred_version, so
```

Comment:
> it increments its current shred_version

There is a slight chance that `current_shred_version + 1 % 0xffff` (shred version is just a `u16`) will equal the new shred_version after the restart. I think this is probably ok as normal gossip shouldn't want to pull/push the restart crds values. Might be worth a mention though

Reply:
The current proposal is just do a CRDS flush upon restart, to avoid sending any trash into the future, so I think we are fine. Mentioned that here.